core/database/foxx/api/user_router.js

```diff
@@ -516,9 +519,47 @@ router.get('/token/set', function(req, res) {
                         refresh: req.queryParams.refresh,
                         expiration: Math.floor(Date.now() / 1000) + req.queryParams.expires_in
                     };
-                    g_db._update(user_id, obj, {
-                        keepNull: false
-                    });
+                    if (token_type && other_token_data) {
+
+                        // NOTE: the edge is only defined for the `u` collection and the `globus_coll` collection.
+                        //  The getUserFromClientID method shows that the user account can also be stored in a UUID collection
+                        //  and in the future support domain accounts. For now, we must ensure that the edge is u to globus_coll
+                        const user_doc = g_db.u.exists(user_id);    // TODO: user doc only necessary for edge key, user_doc._key must not have invalid character '/'
+                                                                    //  which should be kicked out by arango
+                        if (!user_doc) {
+                            throw [g_lib.ERR_INVALID_PARAM, "Error writing to globus_token. No such user in u collection '" + user_id + "'"];
+                        }
+
+                        // find or insert collection
+                        const collection_search_key = other_token_data;
+                        let globus_collection = g_db.globus_coll.exists({_key: collection_search_key});
+                        if (!globus_collection) {
+                            // TODO: what other information should live on the globus_coll document?
+                            globus_collection = g_db.globus_coll.save({_key: collection_search_key, name: "some name we could look up"});
+                        }
+
+                        const token_key = user_doc._key + "_" + globus_collection._key + "_" + token_type;    // TODO: key formatting/convention
+                        const token_doc = {
```

Comment:
Are there any other important fields?

Reply:
Possibly:

```javascript
{
    "scope": "urn:globus:auth:scope:transfer.api.globus.org:all", // Base scope
    "dependent_scopes": [], // Array of dependent collection-specific scopes
    "request_time": Math.floor(Date.now() / 1000), 
    "last_used": Math.floor(Date.now() / 1000),
    "status": "<active/inactive/expired>", // For tracking token validity
    "collection_display_name": globus_collection.name || null // For easier reference   
}

```

Reply:
Great suggestions! I think the display name and scopes will be harder to retrieve, but we can absolutely stub them for now and get the real content later!

Reply:
I would avoid including the collection_display_name in the edge document, as it will add a piece of information that might need to be synchronized with what is in the actual globus_coll document.

The scopes you need should already be available when you requested the token from the web server, if you needed any. A good default can be found by looking in the authz part of the web server.

https://github.com/ORNL/DataFed/blob/b6b555d4fdd2497bd30eb1ff3d328a47f3058f9a/web/datafed-ws.js#L202